torch/csrc/jit/pybind_utils.h

```diff
@@ -136,22 +155,37 @@ inline TypedIValue toTypedIValue(py::handle input) {
     for (auto entry : dict) {
       auto keyInfo = toDictKeyIValue(entry.first);
       auto valInfo = toTypedIValue(entry.second);
-      if (!keyType) {
-        keyType = keyInfo.second;
-        valueType = valInfo.second;
-      } else {
-        auto unifiedKey = unifyTypes(keyType, keyInfo.second);
-        auto unifiedValue = unifyTypes(valueType, valInfo.second);
-        if (!unifiedKey || !unifiedValue) {
-          AT_ERROR("Dictionary inputs to traced functions must have consistent type");
-        }
-        keyType = *unifiedKey;
-        valueType = *unifiedValue;
+      auto unifiedKey = unifyOrInitializeType(keyType, keyInfo.second);
+      auto unifiedValue = unifyOrInitializeType(valueType, valInfo.second);
+      if (!unifiedKey || !unifiedValue) {
+        AT_ERROR("Dictionary inputs to traced functions must have consistent type");
       }
+      keyType = *unifiedKey;
+      valueType = *unifiedValue;
       elems.insert(std::make_pair(keyInfo.first, valInfo.first));
     }
     return TypedIValue(at::ivalue::GenericDict::create(std::move(elems)),
                        DictType::create(keyType, valueType));
+  } else if (PyList_Check(input.ptr())) {
+    auto list = py::cast<py::list>(input);
+    std::vector<IValue> elems;
+    size_t len = py::len(list);
+    if (!len) {
+      AT_ERROR("List trace inputs must have elements");
+    }
+    elems.reserve(len);
+
+    TypePtr listType = nullptr;
+    for (auto elem: list) {
+      TypedIValue typedVal = toTypedIValue(elem);
+      elems.push_back(typedVal.ivalue());
+      auto unify = unifyOrInitializeType(listType, typedVal.type());
+      if (!unify) {
+        AT_ERROR("List inputs to traced functions must have consistent element type");
+      }
+      listType = *unify;
+    }
+    return trySpecializeTensorList(elems, listType);
```

Comment:
I take it this is only called from tracing? It's a little hard to tell form looking at the code which functions call into this, since there are multiple functions named toIValue. 

Could you add a comment that this is only for tracing and the only specialized list types which are accepted are tensors atm ?